bitswap/server/internal/decision/engine.go

```diff
@@ -827,19 +798,91 @@ func (e *Engine) MessageReceived(ctx context.Context, p peer.ID, m bsmsg.BitSwap
 		})
 	}
 
-	// Push entries onto the request queue
-	if len(activeEntries) > 0 {
+	// Push entries onto the request queue and signal network that new work is ready.
+	if len(activeEntries) != 0 {
 		e.peerRequestQueue.PushTasksTruncated(e.maxQueuedWantlistEntriesPerPeer, p, activeEntries...)
 		e.updateMetrics()
+		e.signalNewWork()
 	}
 	return false
 }
 
+// handleOverflow processes incoming wants that could not be addded to the peer
+// ledger without exceeding the peer want limit. These are handled by trying to
+// make room by canceling existing wants for which there is no block. If this
+// does not make sufficient room, then any lower priority wants that have
+// blocks are canceled.
+func (e *Engine) handleOverflow(ctx context.Context, p peer.ID, wants, overflow []bsmsg.Entry) []bsmsg.Entry {
+	existingWants := e.peerLedger.WantlistForPeer(p)
+	// Sort wl and overflow from least to most important.
+	slices.SortFunc(existingWants, func(a, b wl.Entry) int {
+		return cmp.Compare(a.Priority, b.Priority)
+	})
+	slices.SortFunc(overflow, func(a, b bsmsg.Entry) int {
+		return cmp.Compare(a.Entry.Priority, b.Entry.Priority)
+	})
+
+	queuedWantKs := cid.NewSet()
+	for _, entry := range existingWants {
+		queuedWantKs.Add(entry.Cid)
+	}
+	queuedBlockSizes, err := e.bsm.getBlockSizes(ctx, queuedWantKs.Keys())
+	if err != nil {
+		log.Info("aborting overflow processing", err)
+		return wants
+	}
+
+	// Remove entries for blocks that are not present to make room for overflow.
+	var removed []int
+	for i, w := range existingWants {
+		if _, found := queuedBlockSizes[w.Cid]; !found {
+			// Cancel lowest priority dont-have.
+			if e.peerLedger.CancelWant(p, w.Cid) {
+				e.peerRequestQueue.Remove(w.Cid, p)
+			}
+			removed = append(removed, i)
+			// Add highest priority overflow.
+			lastOver := overflow[len(overflow)-1]
+			overflow = overflow[:len(overflow)-1]
+			e.peerLedger.Wants(p, lastOver.Entry, 0)
+			wants = append(wants, lastOver)
+			if len(overflow) == 0 {
+				break
+			}
+		}
+	}
+
+	// Not enough dont-haves removed. Replace existing entries, that are a
+	// lower priority, with overflow entries.
+	var replace int
+	for _, overflowEnt := range overflow {
+		// Do not compare with removed existingWants entry.
+		for len(removed) != 0 && replace == removed[0] {
+			replace++
+			removed = removed[1:]
+		}
+		if overflowEnt.Entry.Priority <= existingWants[replace].Priority {
+			// Everything in existingWants is equal or more improtant, so this
+			// overflow entry cannot replace any existing wants.
+			continue
```

Comment:
The overflow list is sorted in ascending priority. By continuing, this ignores a priority that is lower that anything on the wantlist and moves to the next item in the overflow list that might have a higher priority and therefore be usable to replace an item on the wantlist. Example, skipping 2, 3, 3.
```
wantlist: 1 4 5 5 5 5 5 
overflow: 2 2 3 3 6 7
            \_____^   continue
```

One might think, so why not sort the overflow in descending order so that we can break as soon as current overflow has lower priority than current wantlist?  The reason not to is because then replacement opportunities would be missed by replacing a low priority wantlist item with a high priority overflow item, and denying a lower priority overflow item from replacing that wantlist item. in the example below, the overflow's 2 is ignored.
```
wantlist: 1 4 5 5 5 5 5 
overflow: 7 6 3 3 2 2
          '-'
```

Reply:
Overflow is now ordered most to least important, so it can break out as soon as a more important (higher priority) is seen in the existing want list.